File: dfsAndbfsGraph.js
// Depth-First Search (DFS) for Graphs
function depthFirstSearchGraph(graph, startNode, goalNode) {
    const visited = new Set();
    const path = [];

    function dfs(node) {
        if (visited.has(node)) {
            return false; // Prevent cycles
        }

        visited.add(node);
        path.push(node);

        if (node === goalNode) {
            return true; // Path found
        }

        if (graph[node]) { // Check if the node has neighbors
            for (const neighbor of graph[node]) {
                if (dfs(neighbor)) {
                    return true; // Path found from neighbor
                }
            }
        }

        path.pop(); // Backtrack
        return false; // No path found from this node
    }

    if (!dfs(startNode)) {
        return null; // No path found
    }

    return path;
}

// Breadth-First Search (BFS) for Graphs
function breadthFirstSearchGraph(graph, startNode, goalNode) {
    const queue = [startNode];
    const visited = new Set();
    const cameFrom = {}; // For path reconstruction
    visited.add(startNode);

    while (queue.length > 0) {
        const currentNode = queue.shift();

        if (currentNode === goalNode) {
            // Reconstruct path
            const path = [];
            let node = goalNode;
            while (node) {
                path.unshift(node);
                node = cameFrom[node];
            }
            return path;
        }

        if (graph[currentNode]) { // Check if the node has neighbors
            for (const neighbor of graph[currentNode]) {
                if (!visited.has(neighbor)) {
                    visited.add(neighbor);
                    cameFrom[neighbor] = currentNode;
                    queue.push(neighbor);
                }
            }
        }
    }
    return null; // No path found
}

// Example Usage (Adjacency List Representation)
const graph = {
    'A': ['B', 'C'],
    'B': ['D', 'E'],
    'C': ['F'],
    'D': [],
    'E': ['F'],
    'F': []
};

const startNode = 'A';
const goalNode = 'F';

const dfsPath = depthFirstSearchGraph(graph, startNode, goalNode);
const bfsPath = breadthFirstSearchGraph(graph, startNode, goalNode);

console.log("DFS Path:", dfsPath);
console.log("BFS Path:", bfsPath);

File: dfsAndbfsGraph.py
# Depth-First Search (DFS) for Graphs
def depth_first_search_graph(graph, start_node, goal_node):
    visited = set()
    path = []

    def dfs(node):
        if node in visited:
            return False  # Prevent cycles

        visited.add(node)
        path.append(node)

        if node == goal_node:
            return True  # Path found

        if node in graph:  # Check if the node has neighbors
            for neighbor in graph[node]:
                if dfs(neighbor):
                    return True  # Path found from neighbor

        path.pop()  # Backtrack
        return False  # No path found from this node

    if not dfs(start_node):
        return None  # No path found

    return path

# Breadth-First Search (BFS) for Graphs
def breadth_first_search_graph(graph, start_node, goal_node):
    queue = [start_node]
    visited = set()
    came_from = {}  # For path reconstruction
    visited.add(start_node)

    while queue:
        current_node = queue.pop(0)

        if current_node == goal_node:
            # Reconstruct path
            path = []
            node = goal_node
            while node:
                path.insert(0, node)
                node = came_from.get(node)
            return path

        if current_node in graph:  # Check if the node has neighbors
            for neighbor in graph[current_node]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    came_from[neighbor] = current_node
                    queue.append(neighbor)

    return None  # No path found

# Example Usage (Adjacency List Representation)
graph = {
    'A': ['B', 'C'],
    'B': ['D', 'E'],
    'C': ['F'],
    'D': [],
    'E': ['F'],
    'F': []
}

start_node = 'A'
goal_node = 'F'

dfs_path = depth_first_search_graph(graph, start_node, goal_node)
bfs_path = breadth_first_search_graph(graph, start_node, goal_node)

print("DFS Path:", dfs_path)
print("BFS Path:", bfs_path)